### PR TITLE
[dev/more_models] Memory optimizations

### PR DESCRIPTION
**Pull request by abhinavkulkarni**
Hey @Sakits,

I have added the following changes:

1. In `entry.py`, in `init_empty_weights` context block, replaced `from_pretrained` with `from_config` as `from_config` does not load weights, whereas `from_pretrained` does (even inside the `init_empty_weights` context block).

2. Changed how the model is loaded from a checkpoint and is dispatched to GPU/CPU devices. The problem I was facing was with non-persistent buffers such as [rotary embedding buffers](https://huggingface.co/tiiuae/falcon-7b-instruct/blob/main/modeling_falcon.py#L77) that are not stored in the `state_dict` (since they are non-persistent) and they remained on the `meta` device. I had to take inspiration from [AutoGPTQ repo](https://github.com/PanQiWei/AutoGPTQ/blob/c2c5a74f4bc3c4cbd5124fd27dddba751a45b870/auto_gptq/modeling/_utils.py#L152) and have copied the method verbatim.

3. Falcon 7B and 40B have slightly different architectures, so added changes regarding those.

4. While instantiating `WQLinear`, parameters are placed on the CPU whereas buffers (zero-point quantization scales and zero points) are placed on CUDA. This was causing CUDA OOM error as buffers were getting accumulated in CUDA memory layer after layer as quantization progressed.

With these changes and by specifying appropriate `--max_memory` argument, I was able to quantize and evaluate models such as `tiiuae/falcon-40b-instruct` and `mosaicml/mpt-30b-instruct` on RTX 3060 with 12GB VRAM. With reduced `n_samples` (`64` instead of `128`), I was able to quantize `falcon/tiiuae-40b-instruct` on the same GPU.

Thanks!

### CODE REVIEW

**Comment by abhinavkulkarni** [comment]
Following is the architecture for `falcon-7b`:

```
RWForCausalLM(
  (transformer): RWModel(
    (word_embeddings): Embedding(65024, 4544)
    (h): ModuleList(
      (0-31): 32 x DecoderLayer(
        (input_layernorm): LayerNorm((4544,), eps=1e-05, elementwise_affine=True)
        (self_attention): Attention(
          (maybe_rotary): RotaryEmbedding()
          (query_key_value): Linear(in_features=4544, out_features=4672, bias=False)
          (dense): Linear(in_features=4544, out_features=4544, bias=False)
          (attention_dropout): Dropout(p=0.0, inplace=False)
        )
        (mlp): MLP(
          (dense_h_to_4h): Linear(in_features=4544, out_features=18176, bias=False)
          (act): GELU(approximate='none')
          (dense_4h_to_h): Linear(in_features=18176, out_features=4544, bias=False)
        )
      )
    )
    (ln_f): LayerNorm((4544,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=4544, out_features=65024, bias=False)
)
```

Following is the architecture for `falcon-40b`:

```
RWForCausalLM(
  (transformer): RWModel(
    (word_embeddings): Embedding(65024, 8192)
    (h): ModuleList(
      (0-59): 60 x DecoderLayer(
        (ln_attn): LayerNorm((8192,), eps=1e-05, elementwise_affine=True)
        (ln_mlp): LayerNorm((8192,), eps=1e-05, elementwise_affine=True)
        (self_attention): Attention(
          (maybe_rotary): RotaryEmbedding()
          (query_key_value): Linear(in_features=8192, out_features=9216, bias=False)
          (dense): Linear(in_features=8192, out_features=8192, bias=False)
          (attention_dropout): Dropout(p=0.0, inplace=False)
        )
        (mlp): MLP(
          (dense_h_to_4h): Linear(in_features=8192, out_features=32768, bias=False)
          (act): GELU(approximate='none')
          (dense_4h_to_h): Linear(in_features=32768, out_features=8192, bias=False)
        )
      )
    )
    (ln_f): LayerNorm((8192,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=8192, out_features=65024, bias=False)
)
```

**Comment by abhinavkulkarni** [comment]
@casperbh96: AWQ quantizes the model layer by layer.

The PR has changes wherein, it loads the model in the CPU and then moves only one layer to GPU. Once the search for that layer is done, it moves that layer back to CPU and moves the next layer to GPU.

Once quantization is done, we need to evaluate both the original model as well as the quantized model to figure out how much degradation has happened. So, the PR also uses `accelerate` to evaluate big models on smaller GPU.

More details are here: https://huggingface.co/docs/accelerate/usage_guides/big_modeling

TL;DR: At no point is CPU used for computation, only for offloading.

**Comment by abhinavkulkarni** [comment]
Hey @Sakits,

> When executing the code, I encountered this error: ModuleNotFoundError: No module named 'awq.utils.utils'.

I had mistakenly not included `awq.utils.utils` module. I have included it now.

> It seems like we might need to replace simple_dispatch_model with dispatch_model.

We need to use `simple_dispatch_model` instead of `dispatch_model` because `dispatch_model` does not move buffers (that are not part of `state_dict`) to GPU device, they remain of `meta` device. I had to take inspiration from [AutoGPTQ repo](https://github.com/PanQiWei/AutoGPTQ/blob/c2c5a74f4bc3c4cbd5124fd27dddba751a45b870/auto_gptq/modeling/_utils.py#L152) and have copied the method verbatim.

> could you kindly start a new PR and combine these commits into one?

I have rebased all the commits into one.

PS: I was able to quantize recently released LLaMA 2 with these changes ([link](https://huggingface.co/abhinavkulkarni/meta-llama-Llama-2-7b-chat-hf-w4-g128-awq) and [link](https://huggingface.co/abhinavkulkarni/meta-llama-Llama-2-13b-chat-hf-w4-g128-awq)).

Thanks!
